lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -1523,15 +1547,20 @@ class DedupPass : public DedupBase<DedupPass> {
           // If module has symbol (name) that must be preserved even if unused,
           // skip it. All symbol uses must be supported, which is not true if
           // non-private.
-          if (!module.isPrivate() || !module.canDiscardOnUseEmpty()) {
+          if (!module.isPrivate() ||
+              (!module.canDiscardOnUseEmpty() && !isa<ClassLike>(*module))) {
             return success();
           }
 
-          // Explicitly skip class-like modules.  This is presently unreachable
-          // due to above and current implementation but check anyway as dedup
-          // code does not handle these or object operations.
+          // Minimal ClassLike dedup--only ClassLikes within FModuleOps.
           if (isa<ClassLike>(*module)) {
-            return success();
+            InstanceGraphNode *node = instanceGraph.lookup(module);
+            bool onlyUsedInModules =
+                llvm::all_of(node->uses(), [](InstanceRecord *use) {
+                  return isa<FModuleOp>(*use->getParent()->getModule());
+                });
+            if (!onlyUsedInModules)
+              return success();
```

Comment:
Why do we need this condition? Don't we need to recursively dedup classes?

Reply:
Yeah, this should be changed. I was seeing nested classes dedup when they shouldn't, so I added this little bit of logic until I figured it out. I think the real change we need is for classes to consider port names, since these are actual part of their public API: https://github.com/llvm/circt/blob/a285ab792b5620569d4ebe0a105dafa8df429016/lib/Dialect/FIRRTL/Transforms/Dedup.cpp#L90.